app/src/main/java/com/example/music_app/data/repositories/LoginRepositoryImpl.kt

```diff
@@ -1,50 +1,120 @@
 import kotlinx.coroutines.flow.flowOn
-import kotlinx.coroutines.flow.map
+import java.time.Instant
 import java.util.Base64
 
 private const val AUTH_HEADER_NAME = "Authorization"
 private const val AUTH_PROPERTY = "Basic"
 private const val TYPE_HEADER_NAME = "Content-Type"
 private const val TYPE_HEADER = "application/x-www-form-urlencoded"
-private const val KEY = "access_token"
+
+private const val TOKEN_KEY = "access_token"
+private const val TIME_KEY = "token_receiving_time"
+private const val REFRESH_TOKEN_KEY = "refresh_token"
+
+private const val AUTH_GRANT_TYPE = "authorization_code"
+private const val REFRESH_GRANT_TYPE = "refresh_token"
+
+private const val TOKEN_LIFETIME = 3600
 
 class LoginRepositoryImpl(
-    private val dataStoreManager: DataStoreManager
-): LoginRepository {
-    private val spotifyAPI = AuthService.getInstance()
+    private val dataStoreManager: DataStoreManager,
+    private val spotifyAPI: AuthService = AuthService.getInstance()
+) : LoginRepository {
 
     override suspend fun requestToken(code: String) = flow {
-        val token = spotifyAPI.getToken(createHeaders(), GRANT_TYPE, code, REDIRECT_URL)
-        emit(if (token.accessToken != null) {
-            saveToken(token.accessToken)
-            Ok(token)
-        } else Err(ResponseError()))
+        val token = spotifyAPI.getToken(
+            headers = createHeaders(),
+            grantType = AUTH_GRANT_TYPE,
+            code = code,
+            redirectUri = REDIRECT_URL
+        )
+        emit(
+            if (token.accessToken != null) {
+                saveToken(token)
+                Ok(token)
+            } else Err(ResponseError())
+        )
     }.flowOn(Dispatchers.IO)
 
-    override suspend fun saveToken(token: String) = dataStoreManager.saveString(token, KEY)
+    override suspend fun requestRefreshToken(): Flow<Result<TokenRefreshResponse, ResponseError>> =
+        flow {
+            val refreshToken = dataStoreManager.getString(REFRESH_TOKEN_KEY)
+            val token =
+                spotifyAPI.getRefreshToken(
+                    headers = createHeaders(),
+                    grantType = REFRESH_GRANT_TYPE,
+                    refreshToken = refreshToken
+                )
+            emit(
+                if (token.accessToken != null) {
+                    saveRefreshToken(token)
+                    Ok(token)
+                } else Err(ResponseError())
+            )
+        }
+
+    override suspend fun isAuthorized(): Result<Boolean, Throwable> =
+        runCatching {
+            dataStoreManager.getString(TOKEN_KEY).isNotEmpty()
+        }
 
-    override fun isAuthorized(): Flow<Boolean> =
-        dataStoreManager.getString(KEY).map(String::isNotEmpty)
+
+    override suspend fun isOutdated(): Result<Boolean, Throwable> =
+        runCatching {
+            val recTimeString: String = dataStoreManager.getString(TIME_KEY)
+            if (recTimeString == "") throw Exception(AppErrors.RecTime.error)
```

Comment:
You could use methods like `isEmpty()` or `isNullOrEmpty()` for such case, applies to all occurancies